### ISSUE TITLE
Mixer adapter scaling issue

### ISSUE THREAD

**Issue by atrifan** [istio/istio#11819]
Hello,

I am unable to scale the mixer / my out of process adapter. I can set a bigger replicas number for both (mixer from the istio configuration yaml and my adapter directly from the yaml) the issue is that I don't receive any traffic on 3 out of 4 adapters.

My adapter is out of process, type authorization. 

Istio version used: 1.0.5

Are there any additional things I need to enable? Can you help me with an example yaml for this mixer/ adapter scaling functionality?

Regards,
Alex

**Comment by kyessenov** [comment]
cc @mandarjog @douglas-reid 

This appears to be due to how connections are established:
1. Mixer creates a handler per config update.
2. A handler dials to remote adapter once (a TCP connection to route h2 over).
3. All requests flow over this connection.

This implies that Mixer balances load only on failure/config change. We can try one of the following alternatives to improve the situation:
1. Plug-into gRPC lookaside load balancer extension to LB per request.
2. Use envoy as a middle proxy.

Thoughts?

**Comment by bianpengyuan** [comment]
Looks like the easiest way is to make adapter service headless and use DNS schema for adapter service address (https://medium.com/google-cloud/loadbalancing-grpc-for-kubernetes-cluster-services-3ba9a8d8fc03). This will not require change in mixer or any extra components. It is not as powerful as external load balancer or middle proxy, but it might be OK considering mostly adapter is simple.

**Comment by kyessenov** [comment]
@bianpengyuan Interesting idea. I did some investigation into this:

1. WithBalancerName is marked as experimental in the current gRPC library version. It does appear to work, but we should be cautious just yet.

2. RoundRobin should work fine with A records for pods for headless services. It seems to watch for DNS resolver updates and maintain a pool of connections to them. We also need `dns://`  (https://github.com/grpc/grpc/blob/master/doc/naming.md#name-syntax) to enable DNS resolver.

3. The default DNS poll seems to be 30 minutes in gRPC. We may want something smaller than this.

**Comment by kyessenov** [comment]
For 3: reference https://github.com/grpc/grpc-go/issues/1663

**Comment by kyessenov** [comment]
@atrifan 

Can you please try the experimental mixer build gcr.io/istio-testing/mixer:e8e0310daa61208f030a82dfec0661cd2ca5cc6f and add the following to your OOP:
1. add keepalive max connection age to server parameters:
```go
       _ = grpc.NewServer(grpc.KeepaliveParams(keepalive.ServerParameters{
                MaxConnectionAge: 1 * time.Minute, 
        }))
```
2. make the service for OOP adapter headless

**Comment by atrifan** [comment]
@kyessenov  thanks I will try this tommorow and come back with an input

**Comment by kyessenov** [comment]
@atrifan any feedback you can provide on using gRPC round robin balancer?

**Comment by atrifan** [comment]
Hi sorry for the late response,

I have changed the image in the policy service to what you have provided and now it gives the following:
```
istio-policy
chart: mixer-1.0.6 istio: mixer release: istio
1 / 1
13 minutes
gcr.io/istio-testing/mixer:e8e0310daa61208f030a82dfec0661cd2ca5cc6f
docker.io/istio/proxyv2:1.0.6
Liveness probe failed: Get http://172.17.0.6:9093/version: dial tcp 172.17.0.6:9093: getsockopt: connection refused
Back-off restarting failed container
```

**Comment by atrifan** [comment]
hei @kyessenov any update on this?

**Comment by kyessenov** [comment]
It appears there is an issue with upgrading 1.0.6 mixer to the modified 1.1 mixer in your setup. Are you using Galley? Can you provide the error log from failing mixer container?

**Comment by atrifan** [comment]
I am not using galley. How can I upgrade my istio 1.0.6 to 1.1 so I can use the modified mixer?

**Comment by duderino** [comment]
Warning:  this P0 is getting very close to the 1.1 cutoff.  It needs a plan asap

**Comment by kyessenov** [comment]
@atrifan What does the log say when you try to upgrade from 1.0.6? We can probably fix the missing flag or whatever changed, but it'd help to know what the issue is.

@duderino gRPC LB PR was merged to 1.1 yesterday. The plan is to recommend on relying on headless services + server-side max connection age, since there is no other simple alternative in gRPC library. The upgrade issue is different, and probably needs to be tracked separetely.

**Comment by duderino** [comment]
@kyessenov @douglas-reid what's the priority on the upgrade issue?  Unknown until we get steps from @atrifan to reproduce?

**Comment by douglas-reid** [comment]
@duderino I marked this as a P0 because yesterday, it was decided that load-balancing for gRPC was a requirement for OOP adapters. The PR was then merged. If we file another issue to track the upgrade that's fine.

**Comment by duderino** [comment]
@atrifan please file a new issue for the upgrade and reference this one.

**Comment by kyessenov** [comment]
I went ahead and tried to install 1.0.6 and changed mixer to 1.1.0-rc.2 in istio-policy and istio-telemetry. That works but the liveness probe eventually forces a pod restart. The reason is that there was a change in the port number used for monitoring between 1.0.6 and 1.1.0-rc.2. To fix this, you need to modify one line in 1.0.6 deployment:
```yaml
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /version
            port: 15014 # set to 9093 in 1.0.6
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 1
```

**Comment by atrifan** [comment]
Thanks so much will give it a try
